nni/compression/pytorch/utils/counter.py

```diff
@@ -24,112 +237,102 @@ def count_flops_params(model: nn.Module, input_size, custom_ops=None, verbose=Tr
+    else:
+        x = (torch.zeros(x).to(original_device), )
 
-            m.register_buffer('weight_mask', weight_mask)
-            hook_module_list.append(m)
-        prev_m = m
 
-    flops, params = profile(model, inputs=(inputs, ), custom_ops=custom_mask_ops, verbose=verbose)
-
-
-    for m in hook_module_list:
-        m._buffers.pop("weight_mask")
-    # Remove registerd buffer on the model, and fixed following issue:
-    # https://github.com/Lyken17/pytorch-OpCounter/issues/96
-    for m in model.modules():
-        if 'total_ops' in m._buffers:
-            m._buffers.pop("total_ops")
-        if 'total_params' in m._buffers:
-            m._buffers.pop("total_params")
+    global register_ops, results
+    register_ops = mode_cls[mode]
+    if custom_ops is None:
+        custom_ops = {}
+    register_ops.update(custom_ops)
 
-    return flops, params
+    results = defaultdict(dict)
+    handler_collection = []
 
-def count_convNd_mask(m, x, y):
-    """
-    The forward hook to count FLOPs and Parameters of convolution operation.
-    Parameters
-    ----------
-    m : torch.nn.Module
-        convolution module to calculate the FLOPs and Parameters
-    x : torch.Tensor
-        input data
-    y : torch.Tensor
-        output data
-    """
-    output_channel = y.size()[1]
-    output_size = torch.zeros(y.size()[2:]).numel()
-    kernel_size = torch.zeros(m.weight.size()[2:]).numel()
+    # set leaf module name
+    prev_m = None
+    for name, m in model.named_modules():
+        m_type = type(m)
 
-    bias_flops = 1 if m.bias is not None else 0
+        if m_type in register_ops.keys():
+            setattr(m, "_name", name)
 
-    if m.weight_mask is not None:
-        output_channel = m.weight_mask.sum() // (m.in_channels * kernel_size)
+        if isinstance(prev_m, PrunerModuleWrapper):
+            weight_mask = prev_m.weight_mask
+            m.weight_mask = weight_mask
 
-    total_ops = output_channel * output_size * (m.in_channels // m.groups * kernel_size + bias_flops)
+        prev_m = m
 
-    m.total_ops += torch.DoubleTensor([int(total_ops)])
+    def add_hooks(m_):
+        if len(list(m_.children())) > 0:
+            return
 
+        _handler = m_.register_forward_hook(count_module_flops_params)
+        handler_collection.append(_handler)
 
-def count_linear_mask(m, x, y):
-    """
-    The forward hook to count FLOPs and Parameters of linear transformation.
-    Parameters
-    ----------
-    m : torch.nn.Module
-        linear to calculate the FLOPs and Parameters
-    x : torch.Tensor
-        input data
-    y : torch.Tensor
-        output data
-    """
-    output_channel = y.numel()
+    model.eval()
+    model.apply(add_hooks)
 
-    bias_flops = 1 if m.bias is not None else 0
+    with torch.no_grad():
+        model(*x)
 
-    if m.weight_mask is not None:
-        output_channel = m.weight_mask.sum() // m.in_features
+    # restore origin status
+    for name, m in model.named_modules():
```

Comment:
If you have stored your results into an external list `results`, registering it into attributes of modules is no longer necessary.